### PR TITLE
CircleCI: Add build jobs that run on Linux ARM64

### PR DESCRIPTION
**Pull request by martin-g**
## What changes were proposed in this pull request?

This is the third part of #2531!
It is based on top of #2555, so it should be merged after #2555 !

This PR introduces the build jobs on Linux ARM64 and a new workflow that runs these jobs every Sunday at 1AM UTC.

## How was this patch tested?

Only CircleCI's config.yml has been modified. CircleCI jobs should still work!

### CODE REVIEW

**Review by Yikun — APPROVED** [review]
LGTM, any more progress?

**Comment by Yikun** [comment]
@martin-g `This branch is out-of-date with the base branch`, would you mind try to do a rebase?

**Comment by martin-g** [comment]
@Yikun Thanks for the reminder!
@Harshg999 Could you please review this PR ? It is the last step to enable testing on ARM64 architecture. Thank you!

**Comment by martin-g** [comment]
I've received an email from CircleCI saying `As of June 3, 2022 scheduled workflows will no longer be supported on CircleCI.`
Scheduled workflows are going to be replaced by scheduled pipelines [1] [2].

Scheduled pipelines could be edited only by project admins via CircleCI's REST APIs or via their web UI.

So, now the questions are:
- Do you want to use the deprecated scheduled workflows with this PR until the support for them is dropped ?
- Or I should rework the PR to use normal (non-scheduled) jobs for ARM64 ?
- Or someone from the team will create the scheduled pipeline via REST/UI ?

1. https://circleci.com/blog/using-scheduled-pipelines/
2. https://circleci.com/docs/2.0/scheduled-pipelines/

**Comment by martin-g** [comment]
@Harshg999 or anyone else from the Hue team: Any feedback on this PR ?

**Comment by Harshg999** [comment]
Hi @martin-g, apologies to keeping this PR and you waiting for a long time.
Supporting Hue on ARM64 (and where to test it) requires some discussion with the team and is on our roadmap item. (label tagged)

Thank you for notifying us with the upcoming CircleCI changes mentioned in the above comment! Looking more on this, I think scheduling via REST/UI or having normal jobs for ARM but having no `required` status (so that it doesn't slow down the CI more for PRs) looks like some viable options right now.

But again, we need a bit discussion on this one as the roadmap items goes further.

**Comment by martin-g** [comment]
Thanks for the answer, @Harshg999 !
Please let me know once you have a decision!

**Comment by Harshg999** [comment]
Hi @martin-g! Thanks for waiting!
Can you rework this PR a bit to have these ARM jobs as normal CI checks for `py2.7` and `py3.8`? For Py3, we only support py3.8 as of now (3.9 and 3.10 will come in the future).

We won't be having the `required` status so it wont slow-down the PRs also plus we can see the ARM CI checks for every PR that way instead of proposed weekly schedule earlier.

**Comment by martin-g** [comment]
Hi @Harshg999 !

Thanks for the update!
I will re-work it !

**Comment by martin-g** [comment]
@Harshg999 I have re-worked it! 
Should I remove the `build-py-36-arm64` job ? You said that only py-3.8 is supported but there is still `build-py-3.6` for AMD64.

**Comment by Harshg999** [comment]
Yes @martin-g, for py3 lets have py3.8 only as of now.

File: .circleci/config.yml [hunk 1]
```diff
@@ -296,10 +352,38 @@ workflows:
...
+      - build-py-38:
+          filters:
+            branches:
+              only:
+                - py3-ci
+      - build-py-38-arm64:
+          filters:
+            branches:
+              only:
+                - py3-ci
```
Comment (Harshg999) [inline .circleci/config.yml:379]: Same, lets run it for all branches except master like line 358-360

File: .circleci/config.yml [hunk 2]
```diff
@@ -296,10 +352,38 @@ workflows:
...
             branches:
               only:
                 - /.*ci-commit-master.*/
+  build-python3:
+    jobs:
+      - build-py-38:
+          filters:
+            branches:
+              only:
+                - py3-ci
```
[the remark this answers is not in the document]
Comment (martin-g) [reply .circleci/config.yml]: Done!

File: .circleci/config.yml (hunk 1 above)
Comment (martin-g) [reply .circleci/config.yml]: Done!

**Comment by martin-g** [comment]
> for py3 lets have py3.8 only as of now.

Done! I've removed the py-36 jobs for both AMD64 and ARM64!

**Comment by martin-g** [comment]
@Harshg999 Could you please take a look at https://app.circleci.com/pipelines/github/martin-g/hue/100/workflows/287080f0-db26-4df0-8327-95c71115e2ff/jobs/260 ?

Any idea why it fails with:
```
--- Virtual environment /home/circleci/repo/build/env ready
# Normally dev only do "make apps", but build system is calling "make apps docs"
--- skip installing PIP_MODULES
make[1]: Entering directory '/home/circleci/repo/desktop'
"PYTHON_VER is python3.8."
"SYS_PYTHON is /usr/bin/python3.8."
"ENV_PYTHON is /home/circleci/repo/build/env/bin/python3.8."
"----------------> ENV_INSTALL_TARGETS: .recursive-env-install/core .recursive-env-install/libs/aws .recursive-env-install/libs/azure .recursive-env-install/libs/hadoop .recursive-env-install/libs/indexer .recursive-env-install/libs/libanalyze .recursive-env-install/libs/liboauth .recursive-env-install/libs/liboozie .recursive-env-install/libs/librdbms .recursive-env-install/libs/libsaml .recursive-env-install/libs/libsentry .recursive-env-install/libs/libsolr .recursive-env-install/libs/libzookeeper .recursive-env-install/libs/metadata .recursive-env-install/libs/notebook .recursive-env-install/libs/dashboard .recursive-env-install/libs/kafka"
make -C core env-install
make[2]: Entering directory '/home/circleci/repo/desktop/core'
"PYTHON_VER is python3.8."
"SYS_PYTHON is /usr/bin/python3.8."
"ENV_PYTHON is /home/circleci/repo/build/env/bin/python3.8."
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
--- Installing /home/circleci/repo/desktop/core/requirements.txt into virtual-env via /home/circleci/repo/build/env/bin/python3.8 /home/circleci/repo/build/env/bin/pip
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
make[2]: *** [/home/circleci/repo/Makefile.sdk:119: ext-env-pip-install] Error 127
make[2]: Leaving directory '/home/circleci/repo/desktop/core'
make[1]: *** [Makefile:106: .recursive-env-install/core] Error 2
make[1]: Leaving directory '/home/circleci/repo/desktop'
make: *** [Makefile:157: desktop] Error 2

Exited with code exit status 2

CircleCI received exit code 2
``` 
?

I needed to copy the 'install prerequisites' code from [.github/workflows/commitflow-py3.yml](https://github.com/cloudera/hue/blob/9a708b97aaba803f89e3ba12ccf541bdf16a3d3b/.github/workflows/commitflow-py3.yml#L39-L44)

Now the build on AMD64 passes but fails on ARM64. I guess something more needs to be installed on the Ubuntu ARM64 image that comes pre-installed on the AMD64 image.

**Comment by Harshg999** [comment]
Hum, looking a bit above in the trace, it looks like...

```
W: --force-yes is deprecated, use one of the options starting with --allow instead.
"PYTHON_VER is python3.8."
"SYS_PYTHON is /usr/bin/python3.8."
"ENV_PYTHON is /home/circleci/repo/build/env/bin/python3.8."
--- Creating virtual environment at /home/circleci/repo/build/env
Defaulting to user installation because normal site-packages is not writeable
Requirement already satisfied: pip in /usr/local/lib/python3.8/dist-packages (22.0.4)
Defaulting to user installation because normal site-packages is not writeable
Collecting virtualenv
  Downloading virtualenv-20.14.1-py2.py3-none-any.whl (8.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0.0/8.8 MB ? eta -:--:--     ━━━━━━━━━━━━━━━━━━╸━━━━━━━━━━━━━━━━━━━━━ 4.2/8.8 MB 126.1 MB/s eta 0:00:01     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╸ 8.8/8.8 MB 132.7 MB/s eta 0:00:01     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.8/8.8 MB 88.1 MB/s eta 0:00:00
Requirement already satisfied: six<2,>=1.9.0 in /usr/lib/python3/dist-packages (from virtualenv) (1.14.0)
Collecting filelock<4,>=3.2
  Downloading filelock-3.6.0-py3-none-any.whl (10.0 kB)
Collecting platformdirs<3,>=2
  Downloading platformdirs-2.5.2-py3-none-any.whl (14 kB)
Collecting distlib<1,>=0.3.1
  Downloading distlib-0.3.4-py2.py3-none-any.whl (461 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0.0/461.2 KB ? eta -:--:--     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 461.2/461.2 KB 50.4 MB/s eta 0:00:00
Installing collected packages: distlib, platformdirs, filelock, virtualenv
  WARNING: The script virtualenv is installed in '/home/circleci/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed distlib-0.3.4 filelock-3.6.0 platformdirs-2.5.2 virtualenv-20.14.1
created virtual environment CPython2.7.18.final.0-64 in 460ms
  creator CPython2Posix(dest=/home/circleci/repo/build/env, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, wheel=bundle, setuptools=bundle, via=copy, app_data_dir=/home/circleci/.local/share/virtualenv)
    added seed packages: pip==20.3.3, setuptools==44.1.1, wheel==0.36.2
  activators PythonActivator,CShellActivator,FishActivator,PowerShellActivator,BashActivator
--- Virtual environment /home/circleci/repo/build/env ready
# Normally dev only do "make apps", but build system is calling "make apps docs"
--- skip installing PIP_MODULES
make[1]: Entering directory '/home/circleci/repo/desktop'
"PYTHON_VER is python3.8."
"SYS_PYTHON is /usr/bin/python3.8."
"ENV_PYTHON is /home/circleci/repo/build/env/bin/python3.8."
"----------------> ENV_INSTALL_TARGETS: .recursive-env-install/core .recursive-env-install/libs/aws .recursive-env-install/libs/azure .recursive-env-install/libs/hadoop .recursive-env-install/libs/indexer .recursive-env-install/libs/libanalyze .recursive-env-install/libs/liboauth .recursive-env-install/libs/liboozie .recursive-env-install/libs/librdbms .recursive-env-install/libs/libsaml .recursive-env-install/libs/libsentry .recursive-env-install/libs/libsolr .recursive-env-install/libs/libzookeeper .recursive-env-install/libs/metadata .recursive-env-install/libs/notebook .recursive-env-install/libs/dashboard .recursive-env-install/libs/kafka"
make -C core env-install
make[2]: Entering directory '/home/circleci/repo/desktop/core'
"PYTHON_VER is python3.8."
"SYS_PYTHON is /usr/bin/python3.8."
"ENV_PYTHON is /home/circleci/repo/build/env/bin/python3.8."
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
```

It is creating virtualenv with py2.7? 
```
created virtual environment CPython2.7.18.final.0-64 in 460ms
```
and because of which adds lower pip version and other packages (for py2.7)?
```
seeder FromAppData(download=False, pip=bundle, wheel=bundle, setuptools=bundle, via=copy, app_data_dir=/home/circleci/.local/share/virtualenv)
    added seed packages: pip==20.3.3, setuptools==44.1.1, wheel==0.36.2
```

Also thats why it is not able to find py3.8 @martin-g ?
```
"PYTHON_VER is python3.8."
"SYS_PYTHON is /usr/bin/python3.8."
"ENV_PYTHON is /home/circleci/repo/build/env/bin/python3.8."
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
/bin/bash: /home/circleci/repo/build/env/bin/python3.8: No such file or directory
```

**Comment by martin-g** [comment]
The problem is here - https://github.com/cloudera/hue/blob/c4b866830ffa8eaa946dab786186b933369137c4/Makefile#L140-L141

It calls 
```
$ pip3.8 install virtualenv
Defaulting to user installation because normal site-packages is not writeable
Requirement already satisfied: virtualenv in /home/circleci/.local/lib/python3.8/site-packages (20.14.1)
Requirement already satisfied: platformdirs<3,>=2 in /home/circleci/.local/lib/python3.8/site-packages (from virtualenv) (2.5.2)
Requirement already satisfied: distlib<1,>=0.3.1 in /home/circleci/.local/lib/python3.8/site-packages (from virtualenv) (0.3.4)
Requirement already satisfied: filelock<4,>=3.2 in /home/circleci/.local/lib/python3.8/site-packages (from virtualenv) (3.6.0)
Requirement already satisfied: six<2,>=1.9.0 in /usr/lib/python3/dist-packages (from virtualenv) (1.14.0)

$ virtualenv build/env
created virtual environment CPython2.7.18.final.0-64 in 131ms
  creator CPython2Posix(dest=/home/circleci/repo/build/env, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, wheel=bundle, setuptools=bundle, via=copy, app_data_dir=/home/circleci/.local/share/virtualenv)
    added seed packages: pip==20.3.3, setuptools==44.1.1, wheel==0.36.2
  activators PythonActivator,CShellActivator,FishActivator,PowerShellActivator,BashActivator

$ ls -la build/env/bin
total 9908
drwxrwxr-x 2 circleci circleci     4096 Apr 29 07:21 .
drwxrwxr-x 5 circleci circleci     4096 Apr 29 07:21 ..
-rw-rw-r-- 1 circleci circleci     2375 Apr 29 07:21 activate
-rw-rw-r-- 1 circleci circleci     1437 Apr 29 07:21 activate.csh
-rw-rw-r-- 1 circleci circleci     3068 Apr 29 07:21 activate.fish
-rw-rw-r-- 1 circleci circleci     1751 Apr 29 07:21 activate.ps1
-rw-rw-r-- 1 circleci circleci     1199 Apr 29 07:21 activate_this.py
-rwxrwxr-x 1 circleci circleci      254 Apr 29 07:21 easy_install
-rwxrwxr-x 1 circleci circleci      254 Apr 29 07:21 easy_install2
-rwxrwxr-x 1 circleci circleci      254 Apr 29 07:21 easy_install-2.7
-rwxrwxr-x 1 circleci circleci      254 Apr 29 07:21 easy_install2.7
-rwxrwxr-x 1 circleci circleci      245 Apr 29 07:21 pip
-rwxrwxr-x 1 circleci circleci      245 Apr 29 07:21 pip2
-rwxrwxr-x 1 circleci circleci      245 Apr 29 07:21 pip-2.7
-rwxrwxr-x 1 circleci circleci      245 Apr 29 07:21 pip2.7
-rwxr-xr-x 1 circleci circleci 10064760 Apr 29 07:21 python
lrwxrwxrwx 1 circleci circleci        6 Apr 29 07:21 python2 -> python
lrwxrwxrwx 1 circleci circleci        6 Apr 29 07:21 python2.7 -> python
-rwxrwxr-x 1 circleci circleci      232 Apr 29 07:21 wheel
-rwxrwxr-x 1 circleci circleci      232 Apr 29 07:21 wheel2
-rwxrwxr-x 1 circleci circleci      232 Apr 29 07:21 wheel-2.7
-rwxrwxr-x 1 circleci circleci      232 Apr 29 07:21 wheel2.7

```

**Comment by martin-g** [comment]
I was able to resolve the issues!
I will clean up and squash the commits soon!

**Comment by Harshg999** [comment]
> I was able to resolve the issues!
> I will clean up and squash the commits soon!

Awesome work @martin-g!

**Comment by martin-g** [comment]
Closing in favour of https://github.com/cloudera/hue/pull/2834!
There I copy/pasted the final changes into a single commit.

**Pull request closed without merging**